### ISSUE TITLE
Massive thread creation

### ISSUE THREAD

**Issue by MineTheCube** [EpicEricEE/ShopChest#130]
Hi,

On my server with +1000 shops, it crashes as soon as few players are online, due to too many thread being created.

I looked into your code, and it may come from `de.epiceric.shopchest.utils.ShopUtils` (line 312 and 384):
```java
if (shop.getHologram() != null) {
    shop.getHologram().hidePlayer(player);
}
```

So you are calling Hologram#hidePlayer for every player and for every shop (except the visible ones), and this method is:
```java
public void hidePlayer(final Player p) {
    new BukkitRunnable() {
        @Override
        public void run() {
            // ...                  
```

So, _if I'm correct_, each time task is called, you are basically starting almost as many tasks as shops times players.

**Comment by MineTheCube** [comment]
Also, I think that you should change a few variables:
- Hologram.visible is currently a `List<Player>`, should be a `Set<UUID>`
- ShopItem.visible is currently a `Map<Player, Boolean>`, should be a `Set<UUID>`

Why?
- List → Set: `contains` [should be faster](https://stackoverflow.com/a/32552348) + no duplicate values
- Player → UUID: it is preferable to store an UUID rather than a player
- Map → Set: instead of boolean values that can be true or false, if it is in set = true, else = false

**Comment by MineTheCube** [comment]
Just noticed that in [`ShopItemListener`](https://github.com/EpicEricEE/ShopChest/blob/master/src/main/java/de/epiceric/shopchest/listeners/ShopItemListener.java#L32), in `PlayerQuitEvent` handler you are not removing visibles holograms from player. Is it intended?

**Comment by EpicEricEE** [comment]
> So, if I'm correct, each time task is called, you are basically starting almost as many tasks as shops times players.

This is correct and I know it's not the best solution. Would it be better to just have one asynchronous thread running whose only purpose is to show and hide holograms and items, or is it enough to not execute the task if the hologram was not visible beforehand?

> - Hologram.visible is currently a List\<Player\>, should be a Set\<UUID\>
> - ShopItem.visible is currently a Map\<Player, Boolean\>, should be a Set\<UUID\>

Yes, this definitely makes sense. I don't know why I even used a ``Map<Player, Boolean>`` for that.

> Just noticed that in ShopItemListener, in PlayerQuitEvent handler you are not removing visibles holograms from player. Is it intended?

I had a reason for that, but I don't remember. I think it could be changed.

**Comment by MineTheCube** [comment]
> This is correct and I know it's not the best solution. Would it be better to just have one asynchronous thread running whose only purpose is to show and hide holograms and items, or is it enough to not execute the task if the hologram was not visible beforehand?

The issue here is not the ShopUpdater task, but in `de.epiceric.shopchest.utils.ShopUtils` (line 312 and 384), you are doing this:
```java
if (shop.getHologram() != null) {
    shop.getHologram().hidePlayer(player);
}
```

So, for almost every player, you are calling `Hologram#hidePlayer`, and this method will register a new task, **even if hologram is already hidden**.

I've fixed this in PR #131, where I [check if update is needed](https://github.com/EpicEricEE/ShopChest/pull/131/files#diff-ae51b1e4059911e74e1d6b1bfc900609R154).

**Comment by MineTheCube** [comment]
Well, actually this is closed by #131.

Since [hologram](https://github.com/EpicEricEE/ShopChest/blob/a4517b36578c02c584d6ef1b6f8b269c4277ce32/src/main/java/de/epiceric/shopchest/nms/Hologram.java#L156)/[item](https://github.com/EpicEricEE/ShopChest/blob/a4517b36578c02c584d6ef1b6f8b269c4277ce32/src/main/java/de/epiceric/shopchest/shop/ShopItem.java#L114) changes are checked using "isVisible", packets are now only sent when needed.
